In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pandas import Series,DataFrame
from sklearn.metrics import r2_score
import statsmodels.formula.api as smf
from sklearn import metrics
%matplotlib inline
from sklearn.linear_model import LinearRegression

### 1)	Load the training and testing sets using the read_csv() function, and save them as dataframes with the names pisaTrain and pisaTest. 
How many students are there in the training set?


In [2]:
pisaTrain = pd.read_csv('pisa2009train.csv')
pisaTest = pd.read_csv('pisa2009test.csv')
print('The number of students in training datasets are: ',pisaTrain.shape[0])

('The number of students in training datasets are: ', 3663)


### 2)	Using groupby on pisaTrain, what is the average reading test score of males and females??

In [3]:
pisaTrain.head()

grade  male   raceeth  preschool  expectBachelors  motherHS  \
0     11     1       NaN        NaN              0.0       NaN   
1     11     1     White        0.0              0.0       1.0   
2      9     1     White        1.0              1.0       1.0   
3     10     0     Black        1.0              1.0       0.0   
4     10     1  Hispanic        1.0              0.0       1.0   

   motherBachelors  motherWork  fatherHS  fatherBachelors      ...       \
0              NaN         1.0       NaN              NaN      ...        
1              1.0         1.0       1.0              0.0      ...        
2              1.0         1.0       1.0              NaN      ...        
3              0.0         1.0       1.0              0.0      ...        
4              0.0         1.0       1.0              0.0      ...        

   englishAtHome  computerForSchoolwork  read30MinsADay  \
0            0.0                    1.0             0.0   
1            1.0                    1.0             1.0   
2            1.0                    1.0             0.0   
3            1.0                    1.0             1.0   
4            1.0                    1.0             1.0   

   minutesPerWeekEnglish  studentsInEnglish  schoolHasLibrary  publicSchool  \
0                  225.0                NaN               1.0             1   
1                  450.0               25.0               1.0             1   
2                  250.0               28.0               1.0             1   
3                  200.0               23.0               1.0             1   
4                  250.0               35.0               1.0             1   

   urban  schoolSize  readingScore  
0      1       673.0        476.00  
1      0      1173.0        575.01  
2      0      1233.0        554.81  
3      1      2640.0        458.11  
4      1      1095.0        613.89  

[5 rows x 24 columns]

In [4]:
pisaTrain.columns

Index([u'grade', u'male', u'raceeth', u'preschool', u'expectBachelors',
       u'motherHS', u'motherBachelors', u'motherWork', u'fatherHS',
       u'fatherBachelors', u'fatherWork', u'selfBornUS', u'motherBornUS',
       u'fatherBornUS', u'englishAtHome', u'computerForSchoolwork',
       u'read30MinsADay', u'minutesPerWeekEnglish', u'studentsInEnglish',
       u'schoolHasLibrary', u'publicSchool', u'urban', u'schoolSize',
       u'readingScore'],
      dtype='object')

In [5]:
rscorebymean=pisaTrain.groupby('male').mean()['readingScore']
print('The average reading score of male candidates are : %1.2f' %rscorebymean[1])
print('The average reading score of female candidates are : %1.2f' %rscorebymean[0])

The average reading score of male candidates are : 483.53
The average reading score of female candidates are : 512.94


### 3)	Which variables have missing data in at least one observation in the training set? 

In [6]:
null_columns=pisaTrain.columns[pisaTrain.isnull().any()]
pisaTrain[null_columns].isnull().sum()

raceeth                   35
preschool                 56
expectBachelors           62
motherHS                  97
motherBachelors          397
motherWork                93
fatherHS                 245
fatherBachelors          569
fatherWork               233
selfBornUS                69
motherBornUS              71
fatherBornUS             113
englishAtHome             71
computerForSchoolwork     65
read30MinsADay            34
minutesPerWeekEnglish    186
studentsInEnglish        249
schoolHasLibrary         143
schoolSize               162
dtype: int64

The variables having missing data are as follows: raceeth (35), preschool (56), expectBachelors (62), motherHS (97), motherBachelors (397), motherWork (93), fatherHS (245), fatherBachelors (569), fatherWork (233), selfBornUS (69), motherBornUS (71)), fatherBornUS (113), englishAtHome (71), computerForSchoolwork (65), read30MinsADay (34), minutesPerWeekEnglish (186), studentsInEnglish (249), schoolHasLibrary (143), schoolSize (162)

### 4)	Linear regression discards observations with missing data, so we will remove all such observations from the training and testing sets. Ideally we should be filling the missing values with plausible information, but for now, lets remove them and move ahead. 

Type the following commands into your notebook to remove observations with any missing value from pisaTrain and pisaTest: 

pisaTrain = train.dropna(axis=0) #axis=0 because we want to delete the rows having NA 

pisaTest = test.dropna(axis=0) 

### How many observations are now in the training set? 
### How many observations are now in the testing set?


In [7]:
pisaTrain = pisaTrain.dropna(axis=0)
pisaTest= pisaTest.dropna(axis=0)
print('The number of observation in training dataset after droping missing data are ', pisaTrain.shape[0])
print('The number of observation in testing dataset after droping missing data are: ', pisaTest.shape[0])

('The number of observation in training dataset after droping missing data are ', 2414)
('The number of observation in testing dataset after droping missing data are: ', 990)


### 5)	Which of the following variables is an unordered factor/categorical variable with at least 3 levels? 

In [8]:
pisaTrain.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2414 entries, 1 to 3662
Data columns (total 24 columns):
grade                    2414 non-null int64
male                     2414 non-null int64
raceeth                  2414 non-null object
preschool                2414 non-null float64
expectBachelors          2414 non-null float64
motherHS                 2414 non-null float64
motherBachelors          2414 non-null float64
motherWork               2414 non-null float64
fatherHS                 2414 non-null float64
fatherBachelors          2414 non-null float64
fatherWork               2414 non-null float64
selfBornUS               2414 non-null float64
motherBornUS             2414 non-null float64
fatherBornUS             2414 non-null float64
englishAtHome            2414 non-null float64
computerForSchoolwork    2414 non-null float64
read30MinsADay           2414 non-null float64
minutesPerWeekEnglish    2414 non-null float64
studentsInEnglish        2414 non-null float64
schoo

In [9]:
pisaTrain.raceeth.value_counts()

White                                     1470
Hispanic                                   500
Black                                      228
Asian                                       95
More than one race                          81
American Indian/Alaska Native               20
Native Hawaiian/Other Pacific Islander      20
Name: raceeth, dtype: int64

In [10]:
print('Categorical column(s) with more than 3 levels are: Raceeth')

Categorical column(s) with more than 3 levels are: Raceeth


### 6)	To include unordered categorical variables in a linear regression model, we define one level as the "reference level" and add a binary variable for each of the remaining levels. In this way, a categorical variable with n levels is replaced by n-1 binary variables. The reference level is typically selected to be the most frequently occurring level in the data-set or the first level in alphabetical order. ( We can change from one Level to another as well, Note:- This makes no difference to our solution, it remains the same irrespective of the reference taken) 

As an example, consider the unordered categorical variable variable "color", with levels "red", "green", and "blue". If "green" were the reference level, then we would add binary variables "colorRed" and "colorBlue" to a linear regression problem. All red examples would have colorRed=1 and colorBlue=0. All blue examples would have colorRed=0 and colorBlue=1. All green examples would have colorRed=0 and colorBlue=0. 

Now, consider the variable "raceeth" in our problem, which has levels "American Indian/Alaska Native", "Asian", "Black", "Hispanic", "More than one race", "Native Hawaiian/Other Pacific Islander", and "White". 

### If we select White as the reference level, which binary variables will be included in the regression model? (Select all that apply.)


In [11]:
df_raceeth = pd.get_dummies(pisaTrain['raceeth'])

df_raceeth.drop(labels=['White'],axis=1,inplace=True)
#dropping White as it is the reference

print('Binary variables that will be included in the regression model are..\n')
for i,j in  DataFrame(df_raceeth.columns).iterrows():
    print(DataFrame(df_raceeth.columns)[0][i])

Binary variables that will be included in the regression model are..

American Indian/Alaska Native
Asian
Black
Hispanic
More than one race
Native Hawaiian/Other Pacific Islander


### 7)	Consider again adding our unordered Category race to the regression model with reference level "White". 

### For a student who is Asian, which binary variables would be set to 0? All remaining variables will be set to 1. 


In [12]:
df_Asian = df_raceeth[df_raceeth['Asian']==1]
listColumns = list(df_Asian.columns)

print('Columns that would be set to 0 for an Asian student..\n')
for i in listColumns:
    if(df_Asian[i].all()==0):
        print(i)
#The refernce is white and hence,
print('White')

Columns that would be set to 0 for an Asian student..

American Indian/Alaska Native
Black
Hispanic
More than one race
Native Hawaiian/Other Pacific Islander
White


### 8)	For a student who is white, which binary variables would be set to 0? All remaining variables will be set to 1.

In [13]:
df_Asian = df_raceeth[df_raceeth['Asian']==1]
listColumns = list(df_Asian.columns)

print('Columns that would be set to 0 for a White student..\n')
for i in listColumns:
    print(i)

Columns that would be set to 0 for a White student..

American Indian/Alaska Native
Asian
Black
Hispanic
More than one race
Native Hawaiian/Other Pacific Islander


### 9)	Because the race variable takes on text values, we will have to create Dummy Variables before we run Modeling on the data. 

### Build a linear regression model (call it lmScore) using the training set to predict readingScore using all the remaining variables. (Including the dummies). Remember to omit one of the dummy variable ( the omitted dummy will be the reference) 

### What is the Multiple R-squared value of lmScore on the training set?


In [14]:
pisaTrain = pd.get_dummies(pisaTrain, drop_first= True)
pisaTest = pd.get_dummies(pisaTest, drop_first= True)
x= pisaTrain.drop('readingScore', axis = 1)
y= pisaTrain.readingScore
lmScore = LinearRegression()
lmScore.fit(x,y)
ypred=lmScore.predict(x)
print('R-squared value of LinearRegression model on the training set is',round(r2_score(y,ypred),1))

('R-squared value of LinearRegression model on the training set is', 0.3)


### 10)	What is the training-set root-mean squared error (RMSE) of lmScore?

In [15]:
rmse=round(np.sqrt(metrics.mean_squared_error(y,ypred)),2)
print('Root-Mean Squared Error (RMSE) of lmScore on the training set is',rmse)

('Root-Mean Squared Error (RMSE) of lmScore on the training set is', 73.37)


### 11)	Consider two students A and B. They have all variable values the same, except that student A is in grade 11 and student B is in grade 9. What is the predicted reading score of student A minus the predicted reading score of student B?

In [16]:
coef = lmScore.coef_
difference = round((11*coef[0])-(9*coef[0]),2)
print('The predicted reading score of student A minus the predicted reading score of student B is:',difference)

('The predicted reading score of student A minus the predicted reading score of student B is:', 59.09)


### 12)	What is the meaning of the coefficient associated with variable raceethAsian?

In [17]:
print('The coefficient associated with variable ',x.columns[22],' is : %3.4f' % lmScore.coef_[22])
print('This coefficent means that if the student is asian 63.167 will be added to the reading score ')

('The coefficient associated with variable ', 'raceeth_Asian', ' is : 63.1670')
This coefficent means that if the student is asian 63.167 will be added to the reading score 


### 13)	Based on the significance codes, which variables out of these are candidates for removal from the model? Select all that apply. (The factor variable raceeth should only be removed if none of its levels are significant.)

In [18]:
lm1 = smf.ols('y~x',pisaTrain).fit()
print(lm1.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.325
Model:                            OLS   Adj. R-squared:                  0.317
Method:                 Least Squares   F-statistic:                     41.04
Date:                Tue, 12 Feb 2019   Prob (F-statistic):          1.72e-180
Time:                        04:42:45   Log-Likelihood:                -13795.
No. Observations:                2414   AIC:                         2.765e+04
Df Residuals:                    2385   BIC:                         2.781e+04
Df Model:                          28                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     76.4890     37.303      2.050      0.0

In [19]:
x.columns

Index([u'grade', u'male', u'preschool', u'expectBachelors', u'motherHS',
       u'motherBachelors', u'motherWork', u'fatherHS', u'fatherBachelors',
       u'fatherWork', u'selfBornUS', u'motherBornUS', u'fatherBornUS',
       u'englishAtHome', u'computerForSchoolwork', u'read30MinsADay',
       u'minutesPerWeekEnglish', u'studentsInEnglish', u'schoolHasLibrary',
       u'publicSchool', u'urban', u'schoolSize', u'raceeth_Asian',
       u'raceeth_Black', u'raceeth_Hispanic', u'raceeth_More than one race',
       u'raceeth_Native Hawaiian/Other Pacific Islander', u'raceeth_White'],
      dtype='object')

In [20]:
print('The variables which are the candidates for removal from the model based on significance codes are: \npreschool, \nfatherHS, \nmotherHS, \nmotherWork, \npreschool, \nmotherHS, \nfatherHS, \nselfBornUS, \nmotherBornUS, \nfatherBornUS, \nminutesPerWeekEnglish, \nstudentsInEnglish, \nschoolHasLibrary, \nschoolSize, \nraceeth_Hispanic')

The variables which are the candidates for removal from the model based on significance codes are: 
preschool, 
fatherHS, 
motherHS, 
motherWork, 
preschool, 
motherHS, 
fatherHS, 
selfBornUS, 
motherBornUS, 
fatherBornUS, 
minutesPerWeekEnglish, 
studentsInEnglish, 
schoolHasLibrary, 
schoolSize, 
raceeth_Hispanic


### 14)	Using the "predict" function , use the lmScore model to predict the reading scores of students in pisaTest. Call this list of predictions "predTest". Do not change the variables in the model (for example, do not remove variables that we found were not significant in the previous part of this problem). 

### What is the range between the maximum and minimum predicted reading score on the test set?


In [21]:
x_test = pisaTest.drop('readingScore', axis=1)
y_test = pisaTest.readingScore
predTest = lmScore.predict(x_test)
print('The range of maximum value in pred Test %5.2f' %max(predTest), 'and minimum value %5.2f ' %min(predTest), 'is : %5.0f' %(max(predTest) - min(predTest)))

('The range of maximum value in pred Test 637.69', 'and minimum value 353.22 ', 'is :   284')


### 15)	What is the sum of squared errors (SSE) of lmScore on the testing set? 
_____ (give a integer) 

### What is the root-mean squared error (RMSE) of lmScore on the testing set? 
_____ (give in 2 decimals)


In [22]:
SSE_test= round(np.sum((y_test-predTest)**2))
rmse_test = round(np.sqrt(np.mean((y_test-predTest)**2)),2)
print('The sum of squared errors (SSE) of lmScore on the testing set is : ', int(SSE_test))
print('The root-mean squared error (RMSE) of lmScore on the testing set is : ', rmse_test)

('The sum of squared errors (SSE) of lmScore on the testing set is : ', 5762082)
('The root-mean squared error (RMSE) of lmScore on the testing set is : ', 76.29)


### 16)	What is the predicted test score used in the baseline model? Remember to compute this value using the training set and not the test set.

In [23]:
pred_baseline = np.mean(y)
print('The predicted test score using the baseline model will be : %3.2f' %pred_baseline)

The predicted test score using the baseline model will be : 517.96


### 17)	What is the sum of squared errors of the baseline model on the testing set? HINT: We call the sum of squared errors for the baseline model the total sum of squares (SST).

In [24]:
SSE_baseline= round(np.sum((y_test-pred_baseline)**2))
print('The sum of squared errors of the baseline model on the testing set is ', int(SSE_baseline))

('The sum of squared errors of the baseline model on the testing set is ', 7802354)


### 18)	What is the test-set R-squared value of lmScore?

In [25]:
print('R-squared value of lmScore on the training set is',round(r2_score(y_test,predTest),2))

('R-squared value of lmScore on the training set is', 0.26)
